gapic/src/main/com/google/gapid/views/DeviceDialog.java

```diff
@@ -0,0 +1,322 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.gapid.views;
+
+import static com.google.gapid.util.Logging.throttleLogRpcError;
+import static com.google.gapid.util.MoreFutures.logFailure;
+import static com.google.gapid.widgets.Widgets.createDropDownViewer;
+import static com.google.gapid.widgets.Widgets.createGroup;
+import static com.google.gapid.widgets.Widgets.createLabel;
+import static com.google.gapid.widgets.Widgets.createLink;
+import static com.google.gapid.widgets.Widgets.withIndents;
+import static com.google.gapid.widgets.Widgets.withLayoutData;
+import static java.util.logging.Level.WARNING;
+
+import com.google.gapid.models.Devices;
+import com.google.gapid.models.Devices.DeviceCaptureInfo;
+import com.google.gapid.models.Devices.DeviceValidationResult;
+import com.google.gapid.models.Models;
+import com.google.gapid.proto.device.Device;
+import com.google.gapid.rpc.Rpc;
+import com.google.gapid.rpc.RpcException;
+import com.google.gapid.rpc.SingleInFlight;
+import com.google.gapid.rpc.UiErrorCallback;
+import com.google.gapid.util.Messages;
+import com.google.gapid.util.Paths;
+import com.google.gapid.util.Scheduler;
+import com.google.gapid.util.URLs;
+import com.google.gapid.widgets.DialogBase;
+import com.google.gapid.widgets.LoadingIndicator;
+import com.google.gapid.widgets.Widgets;
+
+import org.eclipse.jface.dialogs.IDialogConstants;
+import org.eclipse.jface.viewers.ArrayContentProvider;
+import org.eclipse.jface.viewers.ComboViewer;
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.jface.viewers.LabelProvider;
+import org.eclipse.jface.viewers.StructuredSelection;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.program.Program;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Event;
+import org.eclipse.swt.widgets.Group;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Link;
+import org.eclipse.swt.widgets.Shell;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+public class DeviceDialog {
+  protected static final Logger LOG = Logger.getLogger(TracerDialog.class.getName());
+
+  private static Object mutex = new Object();
+  private static boolean dialogExists = false;
+
+  private DeviceDialog() {
+  }
+
+  public static void showSelectReplayDeviceDialog(Shell shell, Models models, Widgets widgets) {
+    SelectReplayDeviceDialog dialog = null;
+
+    // This method is called every time the list replay devices are refreshed,
+    // and the dialog it opens may itself refresh the list of replay device.
+    // Thus, make sure to have only one dialog open at a time.
```

Comment:
This sounds like the controller (controller as in MVC) is in the wrong place. The way the UI has been written is that the views contain their own controller that reacts to the model events, since the controller is usually tiny. So to fix this, I think the dialog should just mirror the way this is done in all the other views:

in `onCaptureLoaded` and `onReplayDevicesLoaded`:
```
if (capture and replay devices are loaded) {
  update our UI (device table)
  if (!devices.hasReplayDevice()) {
    if (there's only one compatible device) {
      select it
    } else {
      make ourselves visible
    }
  }
}
```

Reply:
Does this means we create the dialog once, and then it just reacts to events? If so, when/where should this dialog be created? Maybe associate with GraphicsTraceView?

What you describe applies to all views, but I don't think it applies to dialogs, which are typically summoned from another view by some static showFoobarDialog() methods, which is the way I tried to follow here.

Reply:
fixed.